src/v/net/connection_rate.cc

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2022 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "net/connection_rate.h"
+
+#include "seastar/core/coroutine.hh"
+#include "ssx/future-util.h"
+#include "vassert.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/net/inet_address.hh>
+
+#include <algorithm>
+#include <chrono>
+
+namespace net {
+
+connection_rate::connection_rate(
+  const connection_rate_info& rate_info, ss::gate& connection_gate) noexcept
+  : _general_rate(rate_info.max_connection_rate)
+  , _connection_gate(connection_gate) {
+    for (const auto& [ip, max_rate] : rate_info.ovverides) {
+        auto [_, res] = _ovverides.emplace(
+          ip, connection_rate_counter(max_rate));
+        vassert(
+          res, "Can not add ovveride for ip: {}, max_rate: {}", ip, max_rate);
+    }
+}
+
+ss::future<> connection_rate::lock(ss::sstring ip) {
+    auto& sem = find_sem(ip);
+    update_tokens_if_needed(sem);
+    co_await sem.rate.wait(_max_wait_time, 1);
+    spawn_updating_fiber_if_needed(ip, sem);
+}
+
+connection_rate_counter& connection_rate::find_sem(const ss::sstring& ip) {
+    auto ovverides_it = _ovverides.find(ip);
+    if (ovverides_it != _ovverides.end()) {
+        return ovverides_it->second;
+    }
+
+    return _general_rate;
+}
+
+void connection_rate::stop() {
+    _general_rate.rate.broken();
+    for (auto& [_, sem] : _ovverides) {
+        sem.rate.broken();
+    }
+}
+
+void connection_rate::update_tokens_if_needed(
+  connection_rate_counter& rate_counter) {
+    int64_t max_tokens_for_update = rate_counter.max_tokens
```

Comment:
std::max(0, ...) ? 

otherwise you run the chance of having a negative max_tokens? 

This should be a function with a unit test in `rate_counter.next_tokens()` - again tokens is too generic. need a better name.

Reply:
0 <= `rate_counter.rate.current()` <= rate_counter.max_tokens`
`rate_counter.max_tokens` can not be lower 1
But yes, I will do it more safety